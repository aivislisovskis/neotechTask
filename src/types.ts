import { TableData } from './components/table/table.types';

export interface ApiDataRow {
  _id?: string,
  name: string,
  surname: string,
  age: number,
  email?: string,
  company: string,
}

export function ApiDataRowToRowData(data: ApiDataRow[]): TableData {
  return data.reduce((table: TableData, row: ApiDataRow): TableData => {
      return [...table, {
        id: row._id || '',
        data: [
          row.name,
          row.surname,
          String(row.age),
          row.email || '',
          row.company
        ]
      }];
  }, [])
}