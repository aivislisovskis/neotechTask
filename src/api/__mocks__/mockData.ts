import { ApiDataRow } from '../../types';

export const listItems: ApiDataRow[] = [
  {
    _id: 'idofentry1',
    name: 'Andris',
    surname: 'Berzins',
    age: 20,
    email: 'some@email.com',
    company: 'Name Of Company',
  },
  {
    _id: 'idofentry2',
    name: 'Janis',
    surname: 'Berzins',
    age: 20,
    email: 'some@email.com',
    company: 'Name Of Company',
  },
  {
    _id: 'idofentry3',
    name: 'Valis',
    surname: 'Berzins',
    age: 20,
    email: 'some@email.com',
    company: 'Name Of Company',
  },
  {
    _id: 'idofentry4',
    name: 'Zanis',
    surname: 'Berzins',
    age: 20,
    email: 'some@email.com',
    company: 'Name Of Company',
  }
];

export const singleNewItem: ApiDataRow = {
  name: 'Janis',
  surname: 'Berzins',
  age: 44,
  email: 'some@email.com',
  company: 'Name Of Company',
};