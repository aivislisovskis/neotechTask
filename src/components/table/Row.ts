import { RowData, RowOptions } from './table.types';
import { create } from '../../helpers/create';
import { Elements } from '../../helpers/create.types';
import styles from './table.css';
import { Button } from '../../elements/button/Button';

export class Row {
  body: HTMLElement | null = null;
  cells: HTMLElement[] = [];

  constructor(public options: RowOptions) {
    this.body = create(Elements.div, { className: options.header ? styles.headerRow : styles.row, content: this.createCells(options.data) });
    !options.header && this.addActions();
  }

  onEdit = (e: MouseEvent) => {
    this.options.onEdit && this.options.onEdit(this.options.data);
  };

  onDelete = (e: MouseEvent) => {
    this.options.onDelete && this.options.onDelete(this.options.data);
  };

  private addActions() {
    if (this.body) {
      this.body.appendChild(create(Elements.div, {
        className: [
          styles.cell,
          styles.buttonBox
          ],
        content: [
          new Button(this.onEdit, 'Edit').body,
          new Button(this.onDelete, 'Delete').body,
        ]
      }))
    }
  }

  public updateCells(data: RowData) {
    data.data.forEach((text: string, index: number) => {
      this.cells[index].innerHTML = text;
    });
}

  private createCells(data: RowData): Array<HTMLElement> {
      const elements = data.data.map((text: string): HTMLElement => {
        const cell = create(Elements.div, { content: text, className: styles.cell });
        this.cells.push(cell);
        return cell;
      });

    return elements;
  }
}