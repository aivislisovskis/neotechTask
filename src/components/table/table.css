@value size-large, size-medium from "../../styles/media.css";

.body {
    width: 100%;
    display: flex;
    flex-direction: column;
    padding: 10px;
    box-sizing: border-box;
}

.row:nth-child(2n+1) {
    background-color: #eeeeee;
}

.row {
    display: flex;
    flex-direction: column;
    border-bottom: 1px solid;
    justify-content: space-between;
    padding: 5px 5px;
    box-sizing: border-box;
}

.headerRow {
    composes: row;
    display: none;
    font-weight: bold;
    text-transform: capitalize;
    border-bottom: 2px solid #000000;
}

.cell {
    display: flex;
    padding: 5px;
    box-sizing: border-box;
    text-align: center;
    width: 100%;
    overflow: hidden;
    white-space: nowrap;
}

.buttonBox {
   justify-content: flex-end;
}

.addNewRow {
    margin-bottom: 10px;
    display: flex;
    justify-content: flex-end;
}

@media (min-width: size-medium) {
    .row {
        flex-direction: row;
    }

    .headerRow {
        display: flex;
    }

    .cell {
        width: 16.6%;
    }

    .buttonBox {
        max-width: 110px;
    }
}