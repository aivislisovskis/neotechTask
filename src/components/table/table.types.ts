export type TableData = RowData[];
export interface TableOptions {
    data?: TableData,
    headers: string[],
}

export interface RowData {
    id: string | number | null,
    data: string[]
};

export interface RowOptions {
    data: RowData,
    header?: boolean,
}

export function isTableData(data: any): data is TableData {
    return Array.isArray(data) && data.length > 0 && isRowData(data[0]);
}

export function isRowData(data: any): data is RowData {
    return Array.isArray(data) && data.length > 0 && (typeof(data[0]) === 'string');
}