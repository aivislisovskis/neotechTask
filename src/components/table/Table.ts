import { create } from '../../helpers/create';
import { Elements } from '../../helpers/create.types';
import { RowData, TableData, TableOptions } from './table.types';
import styles from './table.css';
import { Row } from './Row';

export class Table {
  body: HTMLElement | null = null;

  constructor(public options: TableOptions) {
    this.body = create(Elements.div, {className: styles.body, content: [this.createHeader({ data: options.headers, id: null }), ...this.prepareInitData()]});
  }

  private createHeader(rowData: RowData): HTMLElement | null {
    if (this.options) {
      const row = new Row({ data: rowData, header: true});
      return row.body;
    }

    return null;
  }

  private prepareInitData(): (HTMLElement | null)[] {
    if (this.options?.data) {
      return this.options.data.map((row: RowData): HTMLElement | null => {
        return this.createRow(row);
      });
    }

    return [];
  }

  private createRow(row: RowData) {
    const rowElement = new Row({ data: row });
    return rowElement.body || null;
  }

  public add(data: TableData) {
      data.forEach((row: RowData) => {
        const rowElement = this.createRow(row);
        rowElement && this.body && this.body.appendChild(rowElement);
      })
  }
}
