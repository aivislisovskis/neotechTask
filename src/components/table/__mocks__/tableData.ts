import { ColumnDefinition, TableData } from '../table.types';
import { ColumnType } from '../../../elements/input/input.types';

export const tableData: TableData = [
  {
    data: ['Janis', 'Berzins', '10', 'mail@mail.ll', 'Some company'],
    id: 'someId',
  },
  {
    data: ['Peteris', 'Klavins', '20', 'rail@mail.ll', 'One company'],
    id: 'someId2',
  },
  {
    data: ['Zane', 'Kocina', '25', 'nail@mail.ll', 'Second company'],
    id: 'someId3',
  },
];

export const tableColumns: ColumnDefinition[] = [
  {
    header: 'name',
    type: ColumnType.string,
  },
  {
    header: 'surname',
    type: ColumnType.string,
  },
  {
    header: 'age',
    type: ColumnType.number,
  },
  {
    header: 'email',
    type: ColumnType.string,
  },
  {
    header: 'company',
    type: ColumnType.string,
  },
  {
    header: 'actions',
    type: ColumnType.button,
  }
];