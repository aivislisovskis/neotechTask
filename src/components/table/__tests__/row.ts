import { Row } from '../Row';
import { RowData, RowOptions } from '../table.types';
import { tableData } from '../__mocks__/tableData';

describe('Table > Row', () => {
  const onEditHandler = jest.fn();
  const onDeleteHandler = jest.fn();

  const rowData: RowOptions = {
    header: false,
    data: tableData[0],
    onEdit: onEditHandler,
    onDelete: onDeleteHandler,
  };

  const mouseEvent = new MouseEvent('click', {
    button: 0,
    buttons: 1,
  });

  it('Simple render', () => {
    const row = new Row(rowData);

    expect(row.cells[0].innerHTML).toBe(rowData.data.data[0]);
    expect(row.cells[1].innerHTML).toBe(rowData.data.data[1]);
    expect(row.cells[2].innerHTML).toBe(rowData.data.data[2]);
    expect(row.cells[3].innerHTML).toBe(rowData.data.data[3]);
    expect(row.cells[4].innerHTML).toBe(rowData.data.data[4]);
  });

  it('*Snapshot', () => {
    const row = new Row(rowData);

    expect(row.body).toMatchSnapshot();
  });

  it('Check Event Handlers', () => {
    const row = new Row(rowData);

    const buttonContainer: ChildNode | undefined = row.body?.childNodes[5];

    if (buttonContainer) {
      buttonContainer.firstChild?.dispatchEvent(mouseEvent)
    }

    expect(onEditHandler).toBeCalled();

    if (buttonContainer) {
      buttonContainer.firstChild?.nextSibling?.dispatchEvent(mouseEvent)
    }

    expect(onDeleteHandler).toBeCalled();
  });

  it('Update cells', () => {
    const row = new Row(rowData);

    const updateData: RowData = {
      data: ['Juris', 'Kalnins', '100', 'some@some.ml', 'Other company'],
      id: 'someId',
    };

    row.updateCells(updateData);

    expect(row.cells[0].innerHTML).toBe(updateData.data[0]);
    expect(row.cells[1].innerHTML).toBe(updateData.data[1]);
    expect(row.cells[2].innerHTML).toBe(updateData.data[2]);
    expect(row.cells[3].innerHTML).toBe(updateData.data[3]);
    expect(row.cells[4].innerHTML).toBe(updateData.data[4]);

    expect(row.body).toMatchSnapshot();
  });

  it('Header row', () => {
    const headerRow = { ...rowData, header: true };
    const row = new Row(headerRow);

    expect(row.body?.childNodes[5]).not.toBeInstanceOf(HTMLElement);
  })
});