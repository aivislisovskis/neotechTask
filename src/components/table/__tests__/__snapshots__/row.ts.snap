// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table > Row *Snapshot 1`] = `
<div>
  <div>
    Janis
  </div>
  <div>
    Berzins
  </div>
  <div>
    10
  </div>
  <div>
    mail@mail.ll
  </div>
  <div>
    Some company
  </div>
  <div
    class=""
  >
    <button>
      Edit
    </button>
    <button>
      Delete
    </button>
  </div>
</div>
`;

exports[`Table > Row Update cells 1`] = `
<div>
  <div>
    Juris
  </div>
  <div>
    Kalnins
  </div>
  <div>
    100
  </div>
  <div>
    some@some.ml
  </div>
  <div>
    Other company
  </div>
  <div
    class=""
  >
    <button>
      Edit
    </button>
    <button>
      Delete
    </button>
  </div>
</div>
`;
